jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java

```diff
@@ -392,83 +358,155 @@ public abstract File getFile()
+            }
+        }
+        else
+        {
+            return false;
+        }
+    }
 
     /**
      * list of resource names contained in the given resource.
      * Ordering is unspecified, so callers may wish to sort the return value to ensure deterministic behavior.
      *
      * @return a list of resource names contained in the given resource, or null.
      * Note: The resource names are not URL encoded.
+     * @deprecated Replace with {@link #getPath()} and {@link Files#walkFileTree(Path, FileVisitor)}.
      */
-    public abstract String[] list();
-
-    /**
-     * Returns the resource contained inside the current resource with the
-     * given name, which may or may not exist.
-     *
-     * @param path The path segment to add, which is not encoded.  The path may be non canonical, but if so then
-     * the resulting Resource will return true from {@link #isAlias()}.
-     * @return the Resource for the resolved path within this Resource, never null
-     * @throws IOException if unable to resolve the path
-     * @throws MalformedURLException if the resolution of the path fails because the input path parameter is malformed, or
-     * a relative path attempts to access above the root resource.
-     */
-    public abstract Resource addPath(String path)
-        throws IOException, MalformedURLException;
-
-    /**
-     * Get a resource from within this resource.
-     */
-    @Override
-    public Resource getResource(String path) throws IOException
+    @Deprecated(forRemoval = true)
+    public String[] list()
     {
-        return addPath(path);
-    }
+        try (DirectoryStream<Path> dir = Files.newDirectoryStream(getPath()))
+        {
+            List<String> entries = new ArrayList<>();
+            for (Path entry : dir)
+            {
+                String name = entry.getFileName().toString();
 
-    // FIXME: this appears to not be used
-    @SuppressWarnings("javadoc")
-    public Object getAssociate()
-    {
-        return _associate;
+                if (Files.isDirectory(entry))
+                {
+                    name += "/";
+                }
+
+                entries.add(name);
+            }
+            int size = entries.size();
+            return entries.toArray(new String[size]);
+        }
+        catch (DirectoryIteratorException e)
+        {
+            LOG.debug("Directory list failure", e);
+        }
+        catch (IOException e)
+        {
+            LOG.debug("Directory list access failure", e);
+        }
+        return null;
     }
 
-    // FIXME: this appear to not be used
-    @SuppressWarnings("javadoc")
-    public void setAssociate(Object o)
+    public Resource resolve(String subPath) throws IOException
     {
-        _associate = o;
+        // Check that the path is within the root,
+        // but use the original path to create the
+        // resource, to preserve aliasing.
+        if (URIUtil.canonicalPath(subPath) == null)
```

Comment:
Ultimately we should have a non allocating method to check this, but no so important as we should rarely have .. segments here.

Reply:
The entire `Resource.resolve()` method is very inefficient as there are many cases where multiple URIs need to be created so that we can eventually call `URI.resolve()` with the same semantics as we used to have.

There are many ways to optimize that, by optimizing how many URIs need to be created, or by using some cheap helper like `URIUtil` or by better aligning our previous `addPath` / `getResource` semantics with the `URI.resolve` ones.

Here's a couple semantic differences between our resolve and URI's:

```
given: "file:/a/b/c" resolve: "d"
 us: "file:/a/b/c/d"
URI: "file:/a/b/d"

given: "file:/a/b/c" resolve: "/d"
 us: "file:/a/b/c/d"
URI: "file:/d"
```

Reply:
Given those differences, best if we call out method add rather than resolve